libcontainer/mount_linux.go

```diff
@@ -121,3 +167,75 @@ func syscallMode(i fs.FileMode) (o uint32) {
 	// No mapping for Go's ModeTemporary (plan9 only).
 	return
 }
+
+// mountFd creates a "mount source fd" (either through open_tree(2) or just
+// open(O_PATH)) based on the provided configuration. This function must be
+// called from within the container's mount namespace.
+//
+// In the case of idmapped mount configurations, the returned mount source will
+// be an open_tree(2) file with MOUNT_ATTR_IDMAP applied. For other
+// bind-mounts, it will be an O_PATH. If the type of mount cannot be handled,
+// the returned mountSource will be nil, indicating that the container init
+// process will need to do an old-fashioned mount(2) themselves.
+//
+// This helper is only intended to be used by goCreateMountSources.
+func mountFd(nsHandles *userns.Handles, m *configs.Mount) (*mountSource, error) {
+	if !m.IsBind() {
+		return nil, errors.New("new mount api: only bind-mounts are supported")
+	}
+	if nsHandles == nil {
+		nsHandles = new(userns.Handles)
+		defer nsHandles.Release()
+	}
+
+	var mountFile *os.File
+	var sourceType mountSourceType
+
+	if m.IsBind() {
+		// We only need to use OPEN_TREE_CLONE in the case where we need to use
+		// mount_setattr(2). We are currently in the container namespace and
+		// thus there is no risk of an opened directory being used to escape
+		// the container. OPEN_TREE_CLONE is more expensive than open(2)
+		// because it requires doing mounts inside a new anonymous mount
+		// namespace.
```

Comment:
Yes, because that's the reason we're not using `OPEN_TREE_CLONE` for everything (I mean, kernel support is another reason but we could have a fallback if we really wanted to use `OPEN_TREE_CLONE` when available).

Reply:
Oh, okay, then can you mention that in the comment? My thought was really: "do we care? Do we care enough to not use it unconditionally? I thought the kernel support was the reason". If we state what you said, that would at least avoid me doing that simplification in the future :-D

Reply:
I mean, that _is_ exactly what the comment is explaining, but I can try to make it a bit clearer.